File: dev/benchmarks/prefill4k_allrows_qmv_c2.cpp
// CPU-only C1/C2 per-output order/byte expectation checks. No Metal backend.
#include "flash/FlashGatheredI8QMV.hpp"
#include <bit>
#include <cmath>
#include <iostream>
#include <string>
#include <vector>

namespace {
namespace qmv = splash::flash::gathered_i8_qmv;
void require(bool value, const char *reason) { if (!value) throw std::runtime_error(reason); }
template<class F> void rejects(F &&f, const char *reason) {
  try { f(); } catch (const std::invalid_argument &) { return; }
  throw std::runtime_error(reason);
}
uint16_t bf16(float value) {
  const uint32_t word = std::bit_cast<uint32_t>(value);
  return uint16_t((word + 0x7fffu + ((word >> 16) & 1u)) >> 16);
}
float number(uint16_t value) { return std::bit_cast<float>(uint32_t(value) << 16); }
float sanitized(uint16_t bits, uint32_t &diagnostics) {
  if ((bits & 0x7f80u) == 0x7f80u) { diagnostics |= 4; return 0.0f; }
  return number(bits);
}
float fixtureTree(std::array<float, 32> values) {
  // This fixed CPU reference tree does not claim Metal's simd_sum tree/order.
  for (uint32_t step = 16; step; step /= 2)
    for (uint32_t lane = 0; lane < step; ++lane) values[lane] += values[lane + step];
  return values[0];
}
std::array<float, 2> scalarC1(const std::vector<uint16_t> &x,
                            const std::array<std::vector<int8_t>, 2> &codes,
                            uint32_t &diagnostics) {
  std::array<float, 2> result{};
  for (uint32_t column = 0; column < 2; ++column) {
    std::array<float, 32> lanes{};
    for (uint32_t lane = 0; lane < 32; ++lane)
      for (uint32_t k = lane; k < x.size(); k += 32)
        lanes[lane] += float(codes[column][k]) * sanitized(x[k], diagnostics);
    result[column] = fixtureTree(lanes);
  }
  return result;
}
std::array<float, 2> sharedC2(const std::vector<uint16_t> &x,
                            const std::array<std::vector<int8_t>, 2> &codes,
                            uint32_t &diagnostics) {
  std::array<std::array<float, 32>, 2> lanes{};
  for (uint32_t lane = 0; lane < 32; ++lane)
    for (uint32_t k = lane; k < x.size(); k += 32) {
      const float activation = sanitized(x[k], diagnostics);
      for (uint32_t column = 0; column < 2; ++column)
        lanes[column][lane] += float(codes[column][k]) * activation;
    }
  return {fixtureTree(lanes[0]), fixtureTree(lanes[1])};
}
void checkColumns() {
  require(qmv::numericalPolicy(1) == qmv::kPolicy && qmv::numericalPolicy(2) == qmv::kC2Policy &&
      qmv::numericalPolicy(1) != qmv::numericalPolicy(2), "C1/C2 reduction identities not separate");
  rejects([] { (void)qmv::numericalPolicy(3); }, "unsupported columns policy admitted");
  for (uint32_t rows : {1u, 2u, 3u, 4u, 8u, 15u, 16u}) {
    const auto g = qmv::geometry(rows);
    require(g.gateColumnGroups / 2 == 80 && g.downColumnGroups / 2 == 320 &&
        (g.gateColumnGroups + g.downColumnGroups) / 2 * rows * 10 == 4000 * rows,
        "C2 exact half-grid launch differs");
  }
  uint64_t pairedColumns = 0;
  for (uint32_t width : {640u, 2560u})
    for (uint32_t fixture = 0; fixture < 16; ++fixture) {
      std::vector<uint16_t> x(width);
      std::array<std::vector<int8_t>, 2> codes{std::vector<int8_t>(width), std::vector<int8_t>(width)};
      for (uint32_t k = 0; k < width; ++k) {
        const int32_t value = int32_t((k * 173 + fixture * 31) % 211) - 105;
        x[k] = bf16(float(value) * (fixture % 2 ? 0.00390625f : 16.0f));
        for (uint32_t column = 0; column < 2; ++column)
          codes[column][k] = int8_t(int32_t((k * 53 + fixture * 73 + column * 31) % 255) - 127);
      }
      if (fixture == 0) { std::fill(x.begin(), x.end(), bf16(-0.0f)); }
      if (fixture == 1) { x[0] = 0x7fc0; x[32] = 0x7f80; x[63] = 0xff80; }
      if (fixture == 2) {
        // Cancellation in each SIMD lane, with different neighboring columns.
        for (uint32_t k = 0; k < width; ++k) {
          x[k] = bf16(k / 32 % 2 ? -128.0f : 128.0f);
          codes[0][k] = 127; codes[1][k] = -126;
        }
        x.back() = bf16(-127.5f);
      }
      uint32_t d1 = 0x80000000u, d2 = d1;
      const auto c1 = scalarC1(x, codes, d1), c2 = sharedC2(x, codes, d2);
      require(d1 == d2 && d1 == (fixture == 1 ? 0x80000004u : 0x80000000u),
          "C1/C2 finite operand/sticky diagnostic expectation differs");
      for (uint32_t column = 0; column < 2; ++column) {
        require(std::bit_cast<uint32_t>(c1[column]) == std::bit_cast<uint32_t>(c2[column]),
            "C2 changed per-output F32 laneK accumulation order");
        for (float scale : {0.003f, 0.125f, 1.0f, 31.75f})
          require(bf16(c1[column] * scale) == bf16(c2[column] * scale),
              "C1/C2 late-scale BF16 byte expectation differs");
        ++pairedColumns;
      }
    }
  const char *oldBase = std::getenv(qmv::kFlag), *oldColumns = std::getenv(qmv::kColumnsFlag);
  const std::string savedBase = oldBase ? oldBase : "", savedColumns = oldColumns ? oldColumns : "";
  const bool hadBase = oldBase != nullptr, hadColumns = oldColumns != nullptr;
  ::unsetenv(qmv::kColumnsFlag); ::setenv(qmv::kFlag, "0", 1);
  require(qmv::requestedColumns() == 1, "C1 default control changed");
  ::setenv(qmv::kColumnsFlag, "2", 1);
  rejects([] { (void)qmv::requestedColumns(); }, "C2 admitted with base QMV flag0");
  ::setenv(qmv::kFlag, "1", 1);
  require(qmv::requestedColumns() == 2, "explicit C2 flag rejected");
  ::setenv(qmv::kColumnsFlag, "1", 1);
  require(qmv::requestedColumns() == 1, "explicit C1 flag rejected");
  for (const char *bad : {"", "0", "3", "true", "02", " 2", "-1"}) {
    ::setenv(qmv::kColumnsFlag, bad, 1);
    rejects([] { (void)qmv::requestedColumns(); }, "malformed columns flag admitted");
  }
  if (hadBase) ::setenv(qmv::kFlag, savedBase.c_str(), 1); else ::unsetenv(qmv::kFlag);
  if (hadColumns) ::setenv(qmv::kColumnsFlag, savedColumns.c_str(), 1); else ::unsetenv(qmv::kColumnsFlag);
  std::cout << "{\"cpu_checks\":\"passed\",\"gpu_work\":false,\"paired_cpu_columns\":" << pairedColumns
            << ",\"paired_gpu_byte_qualification\":\"pending\",\"policy\":\"" << qmv::kC2Policy << "\"}\n";
}
} // namespace
int main(int argc, char **argv) {
  try {
    if (argc != 2 || std::string_view(argv[1]) != "--cpu-self-test")
      throw std::invalid_argument("CPU-only helper requires --cpu-self-test; no GPU mode exists");
    checkColumns(); return 0;
  } catch (const std::exception &error) { std::cerr << error.what() << '\n'; return 1; }
}

File: dev/benchmarks/prefill4k_allrows_qmv_c2.py
#!/usr/bin/env python3
"""Optional C2 private follow-on transform; C1 files remain immutable.

Compose after prefill4k_allrows_qmv.transform and stage extra_files AFTER C1's
extra_files (the augmented header replaces C1's header). No inference on import.
"""
from __future__ import annotations

import argparse
import hashlib
import importlib.util
import json
from pathlib import Path
import sys

ROOT = Path(__file__).resolve().parents[2]
FLAG = "SPLASH_FLASH_ALLROWS_I8_GATHERED_QMV_COLUMNS"
POLICY = "private-gathered-signed-i8-bf16-input-f32-lane32-strided-dot-late-row-scale-bf16-dots-swiglu-sg4-c2-rows1to16-v1"
MARKER = "// Private gathered I8 QMV optional C2 Store overlay v1."
HEADER_RELATIVE = "runtime/flash/FlashGatheredI8QMV.hpp"
METAL_RELATIVE = "runtime/metal/kernels/shared/flash_gathered_i8_qmv_c2.metal"


def c1_module():
    sys.dont_write_bytecode = True
    spec = importlib.util.spec_from_file_location("prefill4k_qmv_c1_component", ROOT / "dev/benchmarks/prefill4k_allrows_qmv.py")
    if spec is None or spec.loader is None:
        raise RuntimeError("cannot import private C1 component")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def replace(text: str, before: str, after: str, count: int = 1) -> str:
    actual = text.count(before)
    if actual != count:
        raise RuntimeError(f"Private QMV C2 source drift: expected{count}, got{actual}: {before!r}")
    return text.replace(before, after)


HEADER = r'''
#ifndef __METAL_VERSION__
namespace splash::flash::gathered_i8_qmv {
inline constexpr const char *kColumnsFlag = "SPLASH_FLASH_ALLROWS_I8_GATHERED_QMV_COLUMNS";
inline constexpr std::string_view kC2Policy =
    "private-gathered-signed-i8-bf16-input-f32-lane32-strided-dot-late-row-scale-bf16-dots-swiglu-sg4-c2-rows1to16-v1";
inline uint32_t requestedColumns() {
  const char *raw = std::getenv(kColumnsFlag);
  uint32_t columns = 1;
  if (raw) {
    const std::string_view value(raw);
    if (value == "1") columns = 1;
    else if (value == "2") columns = 2;
    else throw std::invalid_argument("SPLASH_FLASH_ALLROWS_I8_GATHERED_QMV_COLUMNS must be exactly1 or2");
  }
  if (columns == 2 && !requested())
    throw std::invalid_argument("private QMV columns2 requires SPLASH_FLASH_ALLROWS_I8_GATHERED_QMV=1");
  return columns;
}
inline std::string_view numericalPolicy(uint32_t columns) {
  if (columns == 1) return kPolicy;
  if (columns == 2) return kC2Policy;
  throw std::invalid_argument("private QMV columns policy outside1 or2");
}
} // namespace splash::flash::gathered_i8_qmv
#endif
'''


def extra_files() -> dict[str, str]:
    c1 = c1_module().extra_files()
    return {HEADER_RELATIVE: c1[HEADER_RELATIVE] + HEADER,
            METAL_RELATIVE: (ROOT / "dev/benchmarks/prefill4k_allrows_qmv_c2.metal").read_text()}


def transform(relative: str, text: str) -> str:
    if relative not in ("runtime/flash/FlashInt8ExpertStore.hpp", "runtime/flash/FlashInt8ExpertStore.mm"):
        return text
    if c1_module().MARKER not in text:
        raise RuntimeError("C2 transform requires the C1 Store transform first")
    if MARKER in text:
        raise RuntimeError("private QMV C2 transform already applied")
    text = MARKER + "\n" + text
    if relative.endswith(".hpp"):
        return replace(text, "  [[nodiscard]] bool gatheredQMVEnabled() const;",
            "  [[nodiscard]] bool gatheredQMVEnabled() const;\n  [[nodiscard]] uint32_t gatheredQMVColumns() const;")
    text = replace(text, "  const bool gatheredQMV = gathered_i8_qmv::requested();",
        "  const bool gatheredQMV = gathered_i8_qmv::requested();\n  const uint32_t gatheredColumns = gathered_i8_qmv::requestedColumns();")
    text = replace(text, "std::string(gathered_i8_qmv::kPolicy)",
        "std::string(gathered_i8_qmv::numericalPolicy(gatheredColumns))")
    text = replace(text, "  return impl_->gatheredQMV;", """  if (gathered_i8_qmv::requestedColumns() != impl_->gatheredColumns)
    fail("private gathered I8 QMV columns changed after Store construction");
  return impl_->gatheredQMV;""")
    text = replace(text, "void FlashInt8ExpertStore::addGatheredQMVGateUp(", """uint32_t FlashInt8ExpertStore::gatheredQMVColumns() const {
  (void)gatheredQMVEnabled();
  return impl_->gatheredColumns;
}
void FlashInt8ExpertStore::addGatheredQMVGateUp(""")
    text = replace(text, '  graph.add("flash_gathered_i8_qmv_gate_up_sg4_c1",',
        '  graph.add(impl_->gatheredColumns == 2 ? "flash_gathered_i8_qmv_gate_up_sg4_c2" :\n      "flash_gathered_i8_qmv_gate_up_sg4_c1",')
    text = replace(text, "{g.gateColumnGroups, rows, selections}",
        "{g.gateColumnGroups / impl_->gatheredColumns, rows, selections}")
    text = replace(text, '  graph.add("flash_gathered_i8_qmv_down_sg4_c1",',
        '  graph.add(impl_->gatheredColumns == 2 ? "flash_gathered_i8_qmv_down_sg4_c2" :\n      "flash_gathered_i8_qmv_down_sg4_c1",')
    return replace(text, "{g.downColumnGroups, rows, selections}",
        "{g.downColumnGroups / impl_->gatheredColumns, rows, selections}")


def cpu_self_test(source: Path) -> dict:
    c1 = c1_module()
    for relative in ("runtime/flash/FlashInt8ExpertStore.hpp", "runtime/flash/FlashInt8ExpertStore.mm"):
        original = (source / relative).read_text()
        c1_text = c1.transform(relative, original)
        c2_text = transform(relative, c1_text)
        try:
            transform(relative, c2_text)
        except RuntimeError:
            pass
        else:
            raise AssertionError("C2 duplicate transform admitted")
        if relative.endswith(".mm"):
            for primitive in ("backend.allocateBuffer(", "backend.wrapSharedMemory(", "std::make_shared<Mapping>("):
                if c2_text.count(primitive) != original.count(primitive):
                    raise AssertionError("C2 added mapping/allocation")
            if c2_text.count("/ impl_->gatheredColumns") != 2:
                raise AssertionError("C2 exact half-grid dispatch missing")
            start = original.index("void FlashInt8ExpertStore::addGateUp(")
            end = original.index("} // namespace splash::flash", start)
            if original[start:end] not in c2_text:
                raise AssertionError("original MPP methods changed")
    files = extra_files()
    if POLICY not in files[HEADER_RELATIVE]:
        raise AssertionError("C2 helper/Python policy differs")
    c1_shader = c1.extra_files()[c1.METAL_RELATIVE]
    c2_shader = files[METAL_RELATIVE]
    for shader in (c1_shader, c2_shader):
        if shader.count("k += 32") != 2 or shader.count("#pragma clang fp contract(off)") != 3:
            raise AssertionError("per-output laneK/F32 contract policy differs")
    seed = "splash.private-allrows-target-v1\nsource=s\nstore=t\npolicy=mpp\nmtp=original-trained-bank\n"
    if len({hashlib.sha256((seed + suffix).encode()).hexdigest() for suffix in
            ("", f"small_row_policy={c1.POLICY}\n", f"small_row_policy={POLICY}\n")}) != 3:
        raise AssertionError("MPP/C1/C2 derivative identity collision")
    return {"cpu_source_checks": "passed", "gpu_work": False, "extra_allocations": 0,
            "c1_default_preserved": True, "c2_groups": {"gate": [80, "rows", 10], "down": [320, "rows", 10]},
            "paired_gpu_byte_qualification": "pending"}


def stage(source: Path, destination: Path) -> dict:
    if destination.exists() or source.resolve() == destination.resolve():
        raise ValueError("C2 staging requires a NEW destination")
    destination.mkdir(parents=True)
    c1 = c1_module()
    witness = {}
    for relative in ("runtime/flash/FlashInt8ExpertStore.hpp", "runtime/flash/FlashInt8ExpertStore.mm"):
        before = (source / relative).read_text()
        after = transform(relative, c1.transform(relative, before))
        target = destination / relative
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(after)
        witness[relative] = {"input_sha256": hashlib.sha256(before.encode()).hexdigest(),
                             "output_sha256": hashlib.sha256(after.encode()).hexdigest()}
    files = c1.extra_files() | extra_files()
    for relative, content in files.items():
        target = destination / relative
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(content)
        witness[relative] = {"output_sha256": hashlib.sha256(content.encode()).hexdigest()}
    result = {"c1_policy": c1.POLICY, "c2_policy": POLICY, "flag": c1.FLAG, "columns_flag": FLAG,
              "gpu_work": False, "files": witness}
    (destination / "qmv-overlay-manifest.json").write_text(json.dumps(result, indent=2) + "\n")
    return result


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--source", type=Path, required=True)
    parser.add_argument("--destination", type=Path)
    parser.add_argument("--cpu-self-test", action="store_true")
    args = parser.parse_args()
    if args.cpu_self_test:
        print(json.dumps(cpu_self_test(args.source), indent=2))
    else:
        if args.destination is None:
            parser.error("staging requires --destination")
        print(json.dumps(stage(args.source, args.destination), indent=2))
